core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/part/EnvelopePart.kt

```diff
@@ -0,0 +1,677 @@
+package fr.sncf.osrd.envelope.part
+
+import com.carrotsearch.hppc.DoubleArrayList
+import fr.sncf.osrd.envelope.EnvelopePhysics
+import fr.sncf.osrd.envelope.SearchableEnvelope
+import fr.sncf.osrd.envelope_sim.EnvelopeProfile
+import fr.sncf.osrd.envelope_utils.ExcludeFromGeneratedCodeCoverage
+import fr.sncf.osrd.utils.SelfTypeHolder
+import java.util.*
+import java.util.stream.Collectors
+import kotlin.math.abs
+import kotlin.math.max
+import kotlin.math.min
+
+/**
+ * Envelope parts are polylines over speed and space, defined as a sequence of points.
+ * * single and zero point polylines are not valid
+ * * the position of line points must be strictly increasing
+ * * each segment of the line is thus a step over space, which is indexed from 0 to stepCount()
+ */
+class EnvelopePart(
+    /**
+     * Metadata about this envelope part Attributes are stored and looked up by attribute type. An
+     * envelope part thus cannot have two attributes of the same type
+     */
+    private val attrs: Map<Class<out SelfTypeHolder>, SelfTypeHolder>,
+    /** A list of N spacial offsets */
+    private val positions: DoubleArray,
+    /** A list of N speeds, one per position */
+    private val speeds: DoubleArray,
+    /** A list of N - 1 time deltas between positions */
+    private val timeDeltas: DoubleArray
+) : SearchableEnvelope {
+    // region INTRINSIC DATA FIELDS
+
+    /* !!! These arrays must stay private, as even public final arrays are mutable !!! */
+
+    // endregion
+    // region CACHE FIELDS
+    /**
+     * This property is required for inverse lookups on speeds
+     * https://en.wikipedia.org/wiki/Monotonic_function#Inverse_of_function
+     */
+    private val strictlyMonotonicSpeeds: Boolean = checkStrictlyMonotonic(speeds)
+
+    /* Cache fields must not be public, and must also be lazily computed.
+    This ensures intrinsic data fields can be modified while constructing
+    the envelope part. */
+    /** The highest speed */
+    private var maxSpeedCache = Double.NaN
+
+    /** The smallest speed */
+    private var minSpeedCache = Double.NaN
+
+    /** The time from the start of the envelope, in microseconds. Only read using getTotalTimes. */
+    private var cumulativeUSTimesCache: LongArray? = null
+
+    // endregion
+    // region CONSTRUCTORS
+    /** Creates an EnvelopePart */
+    init {
+        runSanityChecks()
+    }
+
+    /** Creates an EnvelopePart */
+    @ExcludeFromGeneratedCodeCoverage
+    constructor(
+        attrs: Iterable<SelfTypeHolder>,
+        positions: DoubleArray,
+        speeds: DoubleArray,
+        timeDeltas: DoubleArray
+    ) : this(makeAttrs(attrs), positions, speeds, timeDeltas)
+
+    /** Return the given metadata attribute */
+    @Suppress("UNCHECKED_CAST")
+    fun <T : SelfTypeHolder?> getAttr(attrType: Class<T>): T? {
+        return attrs[attrType] as T?
```

Comment:
Separate comment because this one's more important IMO

```suggestion
    fun <T : SelfTypeHolder?> getAttr(attrType: Class<T>): T {
        return attrs[attrType] as T
```

An optional return implies that there's no error if the value isn't in the map. Which could be a possibility, but that's not how it's implemented.

Reply:
Wait wait wait

I forgot that the `[]` operator returns null when it's not in the map. My bad.